src/main/java/org/mockito/internal/creation/instance/ConstructorInstantiator.java

```diff
@@ -47,40 +48,43 @@ public ConstructorInstantiator(Object outerClassInstance) {
         return (T) constructor.newInstance(params);
     }
 
-    private static <T> InstantiationException paramsException(Class<T> cls, Exception cause) {
-        return new InstantiationException(
-                join("Unable to create instance of '" + cls.getSimpleName() + "'.",
-                     "Please ensure that the outer instance has correct type and that the target class has 0-arg constructor."),
-                cause);
-    }
-
-    private static <T> InstantiationException noMatchingConstructor(Class<T> cls) {
-        return new InstantiationException(
-                join("Unable to create instance of '" + cls.getSimpleName() + "'.",
-                     "Unable to find a matching 1-arg constructor for the outer instance.")
-                , null);
+    private InstantiationException paramsException(Class<?> cls, Exception e) {
+        String[] constructorArgTypes = new String[constructorArgs.length];
+        for (int i = 0; i < constructorArgs.length; ++i) {
+            constructorArgTypes[i] = constructorArgs[i].getClass().getName();
+        }
+        return new InstantiationException(join(
+                "Unable to create instance of '" + cls.getSimpleName() + "'.",
+                "Please ensure that it has a constructor that matches these argument types: " + Arrays.toString(constructorArgTypes)) + " and executes cleanly"
+                , e);
     }
 
     private static boolean paramsMatch(Class<?>[] types, Object[] params) {
         if (params.length != types.length) {
             return false;
         }
         for (int i = 0; i < params.length; i++) {
-            if (!types[i].isInstance(params[i])) {
+            if (params[i] == null) {
+                if (types[i].isPrimitive()) {
+                    return false;
+                }
+            } else if ((!types[i].isPrimitive() && !types[i].isInstance(params[i])) ||
+                    (types[i].isPrimitive() && !PRIMITIVES_TO_WRAPPERS.get(types[i]).isInstance(params[i]))) {
                 return false;
             }
         }
         return true;
     }
 
-    private static <T> T noArgConstructor(Class<T> cls) {
-        try {
-            return invokeConstructor(cls.getDeclaredConstructor());
-        } catch (Throwable t) {
-            throw new InstantiationException(join(
-                    "Unable to create instance of '" + cls.getSimpleName() + "'.",
-                    "Please ensure it has 0-arg constructor which invokes cleanly."),
-                                             t);
-        }
+    private static final Map<Class<?>, Class<?>> PRIMITIVES_TO_WRAPPERS = new HashMap<Class<?>, Class<?>>();
```

Comment:
Wasn't aware of it - I'm still finding my way through the source code. Good idea, will do.